prompt2model/param_selector/search_with_optuna.py

```diff
@@ -0,0 +1,245 @@
+                    "weight_decay",
+                    low=hyperparameter_space["min_weight_decay"],
+                    high=hyperparameter_space["max_weight_decay"],
+                ),
+                num_train_epochs=trial.suggest_int(
+                    "num_train_epochs",
+                    low=hyperparameter_space["min_num_train_epochs"],
+                    high=hyperparameter_space["max_num_train_epochs"],
+                ),
+                save_strategy=trial.suggest_categorical(
+                    "save_strategy", hyperparameter_space["save_strategy"]
+                ),
+                per_device_train_batch_size=trial.suggest_categorical(
+                    "per_device_train_batch_size",
+                    hyperparameter_space["per_device_train_batch_size"],
+                ),
+                per_device_eval_batch_size=trial.suggest_categorical(
+                    "per_device_eval_batch_size",
+                    hyperparameter_space["per_device_train_batch_size"],
+                ),
+            )
+            objective_trainer = transformers.Trainer(
+                model=model,
+                args=training_args,
+                data_collator=transformers.DataCollatorForSeq2Seq(
+                    tokenizer=self.generation_model_trainer.tokenizer
+                ),
+                train_dataset=train_dataset,
+                eval_dataset=validation,
+            )
+
+            _ = objective_trainer.train()
+            to_optimize = objective_trainer.evaluate()
+            return to_optimize["eval_loss"]
+
+        study = optuna.create_study(
+            study_name="automatic_hyperparameter_search", direction="minimize"
+        )
+
+        study.optimize(func=objective, n_trials=self.n_trials, gc_after_trial=True)
+        best_hyperparameters = {
+            "learning_rate": float(study.best_params["learning_rate"]),
+            "weight_decay": float(study.best_params["weight_decay"]),
+            "num_train_epochs": int(study.best_params["num_train_epochs"]),
+            "save_strategy": study.best_params["save_strategy"],
+            "per_device_train_batch_size": study.best_params[
+                "per_device_train_batch_size"
+            ],
+        }
+        return best_hyperparameters
+
+    def select_from_hyperparameters(
+        self,
+        training_datasets: list[Dataset],
+        validation: Dataset,
+        hyperparameters: Optional[dict[str, Any]] = None,
+    ) -> tuple[PreTrainedModel, PreTrainedTokenizer]:
+        """Select a model among a set of hyperparameters (given or inferred). # noqa D410
+
+        Args:
+            training_datasets: One or more training datasets for the trainer.
+            validation: A dataset for computing validation metrics.
+            hyperparameters: A dictionary of hyperparameter choices.
+
+        If no hyperparameter_space is choosen, then the default hyperparameter_space
```

Comment:
Suggestion for cleaning up the language here:

```suggestion
        If no hyperparameter_space is specified, then the default hyperparameter_space
```